### PR TITLE
User profile

### PR DESCRIPTION
**Pull request by teodorciuraru**
# Summary
- Adds a complete user profile page that tracks the medicine stock of the user and also information about the user.
- Adds setting the current medicine from the Setup Screen.
- New user profile setup screen and also logic.
- Solves Settings button not working in the `Info Hub` page view's controllers.
- Solves `Plan Trip` button being covered by another view in iPhone 4, iOS 8.4.
- Fast code refactoring.
- Modified `TestsAchievementsManager` because, as for the last patch, it is more encapsulated and the only API to test is getting all the achievements.
# Classes
## `UserProfileViewController`

Has been changed, code and UI, so now it allows the user to update his profile information after pressing the `Edit` button. The changes will be saved, only if they pass our validations, when pressing `Save`.
## `SetupScreenUserProfileViewController`

A new setup screen that appears only when launching the app for the first time. It uses the `User` class to save the data regarding the user after validating it.
## `User`

A NSManagedObject that represents a User entity. It makes sure that there is only one user created and handles functions like creating, updating (`define()`), validating (`validateData()`) and helpful miscellaneous (`isUserAlreadyCreated()`).
## `StringInternalExtensions` and `NSDateInternalExtensions`

Created in order to solve a bug regarding "Redundant conformance to protocol XYZ" when running the app tests. I managed to solve it by splitting the default `StringExtensions` file into itself and `StringInternalExtensions` that is only targeting malaria (not the tests) so that the error message stops appearing. Same for `NSDateExtensions`.
## `LocationTextField` and `GenderTextField`

Two custom views that need almost no setup and can be (and needed to be) easily reused throughout the app (PlanMyTripVC, SetupScreenUserProfileVC, UserProfileVC).

### CODE REVIEW

File: malaria-ios/User.swift [hunk 1]
```diff
@@ -0,0 +1,142 @@
+import Foundation
+import CoreData
+
+class User: NSManagedObject {
+  
+  typealias DefineMethodReturnType = (user: User?, error: UserValidationError?)
```
Comment (bphenriques) [inline malaria-ios/User.swift]: Rename to simpler `MethodReturnType`

File: malaria-ios/ViewControllers/SetupScreen/SetupScreenViewController.swift [hunk 2]
```diff
@@ -4,146 +4,189 @@ import DoneToolbarSwift
...
+    viewContext = CoreDataHelper.sharedInstance.createBackgroundContext()!
+    medicineManager = MedicineManager(context: viewContext)
     
-    private var toolBar: ToolbarWithDone!
+    // Set current stock number.
+    if let medicine = medicineManager.getCurrentMedicine() {
+      let numberOfPills = medicine.remainingMedicine
+      numberOfPillsField.text = String(numberOfPills)
+    } else {
+      numberOfPillsField.text = "0"
```
Comment (bphenriques) [inline malaria-ios/ViewControllers/SetupScreen/SetupScreenViewController.swift:52]: Move "0" to a constant

File: malaria-ios/ViewControllers/SetupScreen/SetupScreenViewController.swift [hunk 3]
```diff
@@ -4,146 +4,189 @@ import DoneToolbarSwift
...
+  @IBAction func sendFeedback(sender: AnyObject) {
+    openUrl(NSURL(string: "mailto:\(PeaceCorpsInfo.mail)"))
+  }
+  
+  @IBAction func doneButtonHandler() {
+    let med = Medicine.Pill(rawValue: medicineNameField.text!)!
+    let remainingPills = Int(self.numberOfPillsField.text!)
+    
+    if remainingPills == 0 {
+      ToastHelper.makeToast("You need to refill your pill stock from the User Profile.", duration: 4)
```
Comment (bphenriques) [inline malaria-ios/ViewControllers/SetupScreen/SetupScreenViewController.swift:112]: ditto. Move to constant the string

File: malaria-ios/ViewControllers/SetupScreen/SetupScreenViewController.swift [hunk 4]
```diff
@@ -4,146 +4,189 @@ import DoneToolbarSwift
...
-        medicineManager.registerNewMedicine(med.name(), interval: med.interval())
-        medicineManager.setCurrentPill(med.name())
-        UserSettingsManager.UserSetting.DidConfiguredMedicine.setBool(true)
-
-        let notificationManager = medicineManager.getCurrentMedicine()!.notificationManager
+    // Avoid showing the alert view if there are no changes
+    if let current = medicineManager.getCurrentMedicine() {
+      
+      // Check if the medicine stock changed.
+      if current.remainingMedicine != Int64(remainingPills!) {
```
Comment (bphenriques) [inline malaria-ios/ViewControllers/SetupScreen/SetupScreenViewController.swift:119]: Is Int64 scattered? If so, create an property on the NSManagedObject :) avoids having to to Int64(value) and related computations that only difficulties the code readbility

File: malaria-ios/ViewControllers/SetupScreenUserProfileViewController.swift [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+    // Try to create the user
+    let userCreationResult = User.define(firstName,
+                                         lastName: lastName,
+                                         age: age,
+                                         email: email,
+                                         gender: gender,
+                                         location: location,
+                                         phone: phone)
+    
+    if let error = userCreationResult.error {
```
Comment (bphenriques) [inline malaria-ios/ViewControllers/SetupScreenUserProfileViewController.swift]: Intead of returning a tuple with the user/error you might use the try catch introduced in earlier swift. It makes more sense here. What do you think?

File: malaria-ios/ViewControllers/UserProfileViewController.swift [hunk 6]
```diff
@@ -230,6 +302,26 @@ extension UserProfileViewController: SaveRemainingPillsProtocol {
...
+
+extension UserProfileViewController: UITextFieldDelegate {
+  
+  func textFieldShouldReturn(textField: UITextField) -> Bool {
+    self.view.endEditing(true)
+    return false
+  }
+  
+  func textFieldDidBeginEditing(textField: UITextField) {
+    let newOffset = CGPointMake(0, textField.frame.origin.y - 30)
```
Comment (bphenriques) [inline malaria-ios/ViewControllers/UserProfileViewController.swift:357]: What does this do?

**Comment by bphenriques** [comment]
Gender should be a UIValuePicker. Btw, some of these values might become optional in the future (I would suggest changing the backend so that it is easier to define which fields are optional and what are mandatory). Age should be a number value picker, location should use the same picker as being used in plan my trip. 

There is no "Done" button on the keyboard to quickly dismiss the keyboard.

The user profile it is better if it is a horizontal paged view that when you press next, it goes to the next screen. As it is, the first screen fades out and the second screen fades in which is not the most fluid experience.

Regarding the rest, everything seems okay! 👍  I would prefer having the updated documentation comments in another PR though (it clutters the PR which is focused on User Profile).

**Comment by bphenriques** [comment]
This never happened to you before (according to previous PR). Why now?
 "Redundant conformance to protocol XYZ"

File: malaria-ios/ViewControllers/UserProfileViewController.swift (hunk 6 above)
Comment (teodorciuraru) [inline malaria-ios/ViewControllers/UserProfileViewController.swift:357]: This one pulls the scrollView up, when selecting a textfield, in order to see what you are typing.

File: malaria-ios/ViewControllers/SetupScreen/SetupScreenViewController.swift (hunk 4 above)
Comment (teodorciuraru) [inline malaria-ios/ViewControllers/SetupScreen/SetupScreenViewController.swift:119]: Sorry! Forgot.

File: malaria-ios/ViewControllers/SetupScreenUserProfileViewController.swift (hunk 5 above)
Comment (teodorciuraru) [inline malaria-ios/ViewControllers/SetupScreenUserProfileViewController.swift]: Best idea ever! Was fun to learn about and code looks great.

File: malaria-ios/NSManagedObjects/User.swift [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+    do {
+      try validateData(firstName,
+                       lastName: lastName,
+                       age: age,
+                       location: location,
+                       email: email,
+                       gender: gender,
+                       phone: phone)
+    }
+    catch let error as UserValidationError {
```
Comment (bphenriques) [inline malaria-ios/NSManagedObjects/User.swift]: you don't need to catch **just** to rethrow. Remove the try catch and add `throws` keyword in the method header

File: malaria-ios/User.swift [hunk 8]
```diff
@@ -34,109 +27,98 @@ class User: NSManagedObject {
...
+      try validateData(firstName,
+                       lastName: lastName,
+                       age: age,
+                       email: email,
+                       gender: gender,
+                       location: location,
+                       phone: phone)
+    }
+    catch let error as UserValidationError {
+      throw error
```
Comment (bphenriques) [inline malaria-ios/User.swift]: ditto

File: malaria-ios/User.swift [hunk 9]
```diff
@@ -3,25 +3,18 @@ import CoreData
 
 class User: NSManagedObject {
```
Comment (bphenriques) [inline malaria-ios/User.swift]: You have replicated User.swift files. Recheck the `git status`command for the list of files that were potentially added/modified. Using `git status` would find the extra `User.swift` file

File: malaria-ios/ViewControllers/GenderTextField.swift [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+      (component: Int, row: Int, object: String) in
+      self.text = object
+    }
+    
+    genderPickerProvider = PickerProvider(selectedCall: didSelectComponentCallback,
+                                          values: genderValues)
+    
+    picker.delegate = genderPickerProvider
+    
+    if text == "" {
```
Comment (bphenriques) [inline malaria-ios/ViewControllers/GenderTextField.swift]: Use stringValue.isEmpty instead

File: malaria-ios/NSManagedObjects/User.swift (hunk 7 above)
Comment (teodorciuraru) [inline malaria-ios/NSManagedObjects/User.swift]: Wow :O

**Pull request merged**